### PR TITLE
Higher computational properties of computational identity types

### PR DESCRIPTION
**Pull request by fredrik-bakke**
- Refactors `ap-binary` to compute as one of the sides in the gray interchange diagram.
- Adds whiskering and horizontal concatenation of Yoneda identifications
- Renames `inv-(yoneda|involutive|computational)-Id` to `inv(ʸ|ⁱ|ʲ)`
- Some touch-ups for the file on the Eckmann-Hilton argument

### CODE REVIEW

**Comment by fredrik-bakke** [comment]
So, turns out it's possible to make horizontal concatenation for any identity type one-sided strictly unital, but I don't know if this is desirable.

**Comment by fredrik-bakke** [comment]
I don't feel particularly inspired to continue working on this atm, so I'm just going to mark it as ready for review.

File: src/foundation/yoneda-identity-types.lagda.md [hunk 1]
```diff
@@ -582,6 +578,8 @@ module _
 
 ### The groupoidal laws for the Yoneda identity types
 
+As we may now observe, associativity and the unit laws holds by `refl`.
```
Comment (fredrik-bakke) [inline src/foundation/yoneda-identity-types.lagda.md:581]:
```suggestion
As we may now observe, associativity and the unit laws hold by `refl`.
```

File: src/foundation/yoneda-identity-types.lagda.md [hunk 2]
```diff
@@ -688,23 +691,62 @@ module _
...
+  ap (r ↦ f r y) p |                    | ap (r ↦ f r y') p
+                   |                    |
+                   ∨                    ∨
+                 f x' y ------------> f x' y'
+                         ap (f x') q
+```
+
+and the fact that the concatenation operation on Yoneda identifications is
+two-sided strictly unital, we obtain an action of binary functions on Yoneda
+identifications that computes on both arguments.
```
Comment (fredrik-bakke) [inline src/foundation/yoneda-identity-types.lagda.md:727]:
````suggestion
We obtain an action of binary functions on Yoneda identifications that computes
on both arguments using one of the two sides in the Gray interchange diagram

```text
                         ap (f x) q
                 f x y -------------> f x y'
                   |                    |
                   |                    |
  ap (r ↦ f r y) p |                    | ap (r ↦ f r y') p
                   |                    |
                   ∨                    ∨
                 f x' y ------------> f x' y'
                         ap (f x') q
```

and the fact that the concatenation operation on Yoneda identifications is
two-sided strictly unital.
````

**Comment by fredrik-bakke** [comment]
Apologies, I forgot to mark this PR ready for review. Would it be possible to get it merged before 13:00 on Monday? I'd like to use the webpage for my supervisor meeting.

File: src/foundation/action-on-identifications-binary-functions.lagda.md [hunk 3]
```diff
@@ -28,6 +28,23 @@ Given a binary operation `f : A → B → C` and
 we call this the
 {{#concept "binary action on identifications of binary functions" Agda=ap-binary}}.
 
+There are a few different ways we can define `ap-binary`. We could define it by
+pattern matching on both `p` and `q`, but this leads to restricted computational
+behaviour. Instead, we define it as the upper concatenation in the Gray
+interchange diagram
```
Comment (EgbertRijke) [inline src/foundation/action-on-identifications-binary-functions.lagda.md:34]: If this is called the Gray interchange diagram, then that probably has a reference

Comment (fredrik-bakke) [reply src/foundation/action-on-identifications-binary-functions.lagda.md:34]: I've briefly looked for a good reference and couldn't find anything. I have it from the following excerpt, but it hardly feels appropriate to use it as a reference:

image

File: src/foundation/path-algebra.lagda.md [hunk 4]
```diff
@@ -284,33 +282,47 @@ interchange-Id² :
...
+  ( horizontal-concat-Id² β γ) ＝
+  ( vertical-concat-Id² (left-whisker-concat p γ) (right-whisker-concat β v))
+inner-interchange-Id² {u = refl} β refl = compute-right-horizontal-concat-Id² β
+
+outer-interchange-Id² :
+  {l : Level} {A : UU l} {x y z : A} {p q : x ＝ y} {u w : y ＝ z}
+  (α : p ＝ q) (δ : u ＝ w) →
+  ( horizontal-concat-Id² α δ) ＝
+  ( vertical-concat-Id² (right-whisker-concat α u) (left-whisker-concat q δ))
+outer-interchange-Id² {p = refl} refl δ = compute-left-horizontal-concat-Id² δ
```
Comment (EgbertRijke) [inline src/foundation/path-algebra.lagda.md:299]: What do these operations interchange? They don't really seem to be interchange laws to me

Comment (fredrik-bakke) [reply src/foundation/path-algebra.lagda.md:299]: They're instances of the general interchange law where the outer or inner identifications are constant

File: src/foundation/path-algebra.lagda.md [hunk 5]
```diff
@@ -174,17 +174,15 @@ right-unit-law-vertical-concat-Id² :
...
+compute-left-horizontal-concat-Id² = left-unit-ap-binary (_∙_)
 
-right-unit-law-horizontal-concat-Id² :
+compute-right-horizontal-concat-Id² :
   {l : Level} {A : UU l} {x y z : A} {p q : x ＝ y} (α : p ＝ q) {u : y ＝ z} →
-  horizontal-concat-Id² α (refl {x = u}) ＝
-  right-whisker-concat α u
-right-unit-law-horizontal-concat-Id² = right-unit-ap-binary (_∙_)
+  horizontal-concat-Id² α refl ＝ right-whisker-concat α u
+compute-right-horizontal-concat-Id² = right-unit-ap-binary (_∙_)
```
Comment (EgbertRijke) [inline src/foundation/path-algebra.lagda.md:185]: Hm.. with "left/right unit law it was at least clear that we were putting `refl` in the left and right positions, but with `compute-left` and `compute-right` that isn't clear anymore. On the other hand, I see that they were not really unit laws in the usual sense of the word. Maybe `compute-left-refl` and `compute-right-refl` would be more descriptive name parts than simply `compute-left` and `compute-right`

File: src/foundation/path-algebra.lagda.md (hunk 4 above)
Comment (fredrik-bakke) [reply src/foundation/path-algebra.lagda.md:299]: I.e., they are degenerate cases of the interchange law

File: src/foundation/yoneda-identity-types.lagda.md [hunk 6]
```diff
@@ -688,23 +691,62 @@ module _
...
-    {x y : A} (p : x ＝ʸ y) → ap-yoneda-Id id p ＝ p
+  compute-ap-id-yoneda-Id : {x y : A} (p : x ＝ʸ y) → ap-yoneda-Id id p ＝ p
   compute-ap-id-yoneda-Id {x} p =
     eq-multivariable-htpy 2
       ( λ _ q →
         ( ap (q ∙ᵣ_) (ap-id (p x refl))) ∙
         ( inv (commutative-preconcatr-refl-Id-yoneda-Id p q)))
 ```
 
+### Action of binary functions on Yoneda identifications
```
Comment (EgbertRijke) [inline src/foundation/yoneda-identity-types.lagda.md:709]: This file is definitely starting to contain way too much material. In a subsequent pull request I propose that we create files

`action-on-yoneda-identifications-functions` and similar for other subjects that can be easily factored out.

**Review by EgbertRijke — APPROVED** [review]

File: src/foundation/path-algebra.lagda.md (hunk 4 above)
Comment (EgbertRijke) [reply src/foundation/path-algebra.lagda.md:299]: Both the naming and the specification don't make that apparent.

Comment (fredrik-bakke) [reply src/foundation/path-algebra.lagda.md]: I'll have to fix it later. Thanks for pointing it out

**Pull request merged**